### ISSUE TITLE
The bpm metadata is not written to the media

### ISSUE THREAD

**Issue by doctorfree** [adamjakab/BeetsPluginXtractor#17]
### Problem

I cannot seem to get the `xtractor` plugin to write the bpm metadata tag to the media file itself. If the file has no `bpm` tag then none is added and if it has an existing bpm tag that is not updated. Is the `write` configuration and command-line option implemented? If so, what am I doing wrong?

The calculated bpm value is correctly being added to the Beets library for the media but not to the media.

Running your command in verbose (`-vv`) mode:

```sh
$ beet -vv xtractor -w overdrive
```

Led to this problem:

```
user configuration: /home/ronnie/.config/beets/config.yaml
data directory: /home/ronnie/.config/beets
plugin paths: 
lastgenre: Loading canonicalization tree ~/.config/beets/genres/genres-tree.yaml
Sending event: pluginload
library database: /home/ronnie/.config/beets/musiclibrary.db
library directory: /home/ronnie/Music
Sending event: library_opened
xtractor: Combined query: AndQuery([AndQuery([AnyFieldQuery('overdrive', ('artist', 'title', 'comments', 'album', 'albumartist', 'genre'), SubstringQuery)]), OrQuery([MatchQuery('bpm', None, True), MatchQuery('average_loudness', None, False), MatchQuery('danceable', None, False), MatchQuery('gender', None, False), MatchQuery('genre_rosamerica', None, False), MatchQuery('voice_instrumental', None, False), MatchQuery('mood_acoustic', None, False), MatchQuery('mood_aggressive', None, False), MatchQuery('mood_electronic', None, False), MatchQuery('mood_happy', None, False), MatchQuery('mood_sad', None, False), MatchQuery('mood_party', None, False), MatchQuery('mood_relaxed', None, False), MatchQuery('mood_mirex', None, False)])])
xtractor: Number of items to be processed: 1
xtractor: Running analysis for: /home/ronnie/Music/Matrika/Uppbeat/00 - Overdrive.mp3
xtractor: Extractor: /usr/bin/essentia_streaming_extractor_music
xtractor: Input: /home/ronnie/Music/Matrika/Uppbeat/00 - Overdrive.mp3
xtractor: Output: /home/ronnie/.config/beets/xtraction_data/2a23b2e4ca2d793afee720d866ff953a.json
xtractor: Profile: /home/ronnie/.config/beets/xtraction_data/profile.yml
xtractor: The process exited with code: 0
xtractor: Process stdout: 
xtractor: Process stderr: [   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/danceability.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/gender.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/genre_dortmund.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/genre_electronic.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/genre_rosamerica.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/genre_tzanetakis.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/ismir04_rhythm.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/mood_acoustic.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/mood_aggressive.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/mood_electronic.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/mood_happy.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/mood_party.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/mood_relaxed.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/mood_sad.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/moods_mirex.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/timbre.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/tonal_atonal.history
[   INFO   ] MusicExtractorSVM: adding SVM model /usr/share/musicplayerplus/svm_models/voice_instrumental.history
[   INFO   ] MusicExtractor: Read metadata
[   INFO   ] MusicExtractor: Compute md5 audio hash, codec, length, and EBU 128 loudness
[   INFO   ] MusicExtractor: Replay gain
[   INFO   ] MusicExtractor: Chromaprint
[   INFO   ] MusicExtractor: Compute audio features
[   INFO   ] MusicExtractor: Compute aggregation
[   INFO   ] MusicExtractor: SVM models
[   INFO   ] All done
Writing results to file /home/ronnie/.config/beets/xtraction_data/2a23b2e4ca2d793afee720d866ff953a.json


xtractor: Audiodata: {'bpm': 110, 'danceability': 1.25136184692, 'beats_count': 337, 'average_loudness': 0.964244484901, 'danceable': 0.999999880791, 'gender': 'male', 'is_male': 0.76497811079, 'is_female': 0.235021904111, 'genre_rosamerica': 'roc', 'voice_instrumental': 'voice', 'is_voice': 0.86198425293, 'is_instrumental': 0.138015776873, 'mood_acoustic': 0.0520507767797, 'mood_aggressive': 0.997299790382, 'mood_electronic': 0.88177973032, 'mood_happy': 0.0596570186317, 'mood_sad': 0.992612183094, 'mood_party': 0.986634969711, 'mood_relaxed': 0.951368510723, 'mood_mirex': 'Cluster5', 'mood_mirex_cluster_1': 0.285546392202, 'mood_mirex_cluster_2': 0.0856884494424, 'mood_mirex_cluster_3': 0.0384233407676, 'mood_mirex_cluster_4': 0.134965896606, 'mood_mirex_cluster_5': 0.455375909805}
Sending event: database_change
Sending event: cli_exit
mpdupdate: Updating MPD database...
mpdupdate: Database updated.
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/1970s.m3u
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/ween.m3u
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/1990s.m3u
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/1980s.m3u
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/2020s.m3u
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/2000s.m3u
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/2010s.m3u
playlist: Updating playlist: /home/ronnie/.config/mpd/playlists/ReleasedIn1971.m3u
```
The Beets library was correctly updated with the calculated bpm for this song but the media file itself still has no bpm tag:
```
beet info -l overdrive | grep bpm
 bpm: 110
```

```
beet info overdrive | grep bpm
```

### Setup

* OS: Ubuntu Linux 20.04
* Python version: 3.8.10
* Beets version: 1.6.0
* Turning off other plugins made problem go away (yes/no): no

My configuration (output of `beet config`) is:

```yaml
directory: ~/Music
library: ~/.config/beets/musiclibrary.db

lyrics:
    bing_lang_from: []
    auto: yes
    sources: musixmatch genius
    bing_client_secret: REDACTED
    bing_lang_to:
    google_API_key: REDACTED
    google_engine_ID: REDACTED
    genius_api_key: REDACTED
    fallback:
    force: no
    local: no

import:
    copy: yes
    move: yes
    write: yes
    incremental: yes
    incremental_skip_later: yes
    quiet: no
    quiet_fallback: asis
    resume: yes
    from_scratch: no
    default_action: apply
    detail: yes
    non_rec_action: ask
    duplicate_action: skip
    group_albums: no
    autotag: yes
    bell: no
    log: ~/.config/beets/import.log

plugins:
    - acousticbrainz
    - albumtypes
    - bandcamp
    - describe
    - duplicates
    - edit
    - extrafiles
    - fromfilename
    - hook
    - importadded
    - info
    - lyrics
    - lastgenre
    - missing
    - mbsync
    - mpdstats
    - play
    - playlist
    - smartplaylist
    - mpdupdate
    - unimported
    - xtractor

acousticbrainz:
    auto: yes
    force: no
    tags: []

albumtypes:
    types:
        -   ep: EP
        -   single: Single
        -   soundtrack: OST
        -   live: Live
        -   compilation: Anthology
        -   remix: Remix
    ignore_va: compilation
    bracket: '[]'

bandcamp:
    include_digital_only_tracks: yes
    search_max: 2
    art: no
    comments_separator: "\n---\n"
    exclude_extra_fields: [comments, lyrics]
    genre:
        capitalize: no
        maximum: 0
        always_include: []
        mode: progressive

duplicates:
    album: yes
    checksum: ''
    copy: ''
    count: no
    delete: no
    format: ''
    full: no
    keys: []
    merge: no
    move: ''
    path: no
    tiebreak: {}
    strict: no
    tag: ''

extrafiles:
    patterns:
        all: '*.*'
    paths: {}

mpdstats:
    host: localhost
    port: 6600
    rating: yes
    rating_mix: 0.75

lastgenre:
    auto: yes
    canonical: ~/.config/beets/genres/genres-tree.yaml
    count: 1
    fallback: Unknown
    min_weight: 10
    prefer_specific: yes
    source: track
    separator: ', '
    whitelist: ~/.config/beets/genres/genres.txt
    title_case: yes
    force: yes

play:
    command: mpc --quiet $args load
    raw: yes
    use_folders: no
    relative_to:
    warning_threshold: 100
    bom: no

playlist:
    auto: yes
    playlist_dir: ~/.config/mpd/playlists
    relative_to: ~/Music
    forward_slash: no

smartplaylist:
    relative_to: ~/Music
    playlist_dir: ~/.config/mpd/playlists
    forward_slash: no
    playlists:
        -   name: rating.m3u
            query: rating:0.4..1.0
        -   name: ween.m3u
            query: artist:Ween
        -   name: 1940s.m3u
            query: year:1940..1949
        -   name: 1950s.m3u
            query: year:1950..1959
        -   name: 1960s.m3u
            query: year:1960..1969
        -   name: 1970s.m3u
            query: year:1970..1979
        -   name: 1980s.m3u
            query: year:1980..1989
        -   name: 1990s.m3u
            query: year:1990..1999
        -   name: 2000s.m3u
            query: year:2000..2009
        -   name: 2010s.m3u
            query: year:2010..2019
        -   name: 2020s.m3u
            query: year:2020..2029
        -   name: ReleasedIn$year.m3u
            query: year::197(0|1|2|3|4|5|6|7|8|9)
    auto: yes
    prefix: REDACTED
    urlencode: no

mpd:
    host: localhost
    port: 6600
    rating: yes
    rating_mix: 0.75
    password: REDACTED
    music_directory: ~/Music
    strip_path: ''

musicbrainz:
    extra_tags:
        - year
        - catalognum
        - country
        - media
        - label

match:
    strong_rec_thresh: 0.075
    medium_rec_thresh: 0.125
    max_rec:
        unmatched_tracks: low
        missing_tracks: strong
    distance_weights:
        album: 6.0
        artist: 7.0
        totaltracks: 4.0
        missing_tracks: 1.0
        unmatched_tracks: 10.0
    preferred:
        countries: [US, XW, GB|UK]
        media: [Digital Media|File, CD]
    ignored_media:
        - DVD
        - DVD-Video
        - Blu-ray
        - HD-DVD
        - VCD
        - SVCD
        - UMD
        - VHS

paths:
    default: '%if{$albumartist,%title{$albumartist},%if{$artist,%title{$artist},Unknown Artist}}/%if{$album,$album%aunique{},Unknown Album}%if{$atypes, $atypes}/$track - $title'
    singleton: '%if{$albumartist,%title{$albumartist},%if{$artist,%title{$artist},Unknown Artist}}/%if{$album,$album%aunique{},Unknown Album}%if{$atypes, $atypes}/$track - $title'
    comp: Compilations/%if{$album,$album%aunique{},Unknown Album}%if{$atypes, $atypes}/$track - $artist - $title
    albumtype:soundtrack: Soundtracks/%if{$album,$album%aunique{},Unknown Album}%if{$atypes, $atypes}/$track - $title

importadded:
    preserve_mtimes: yes
    preserve_write_mtimes: yes

threaded: yes

unimported:
    ignore_extensions: jpg png
    ignore_subdirectories: data tmp

xtractor:
    auto: no
    dry-run: no
    write: yes
    threads: 1
    force: no
    quiet: no
    keep_output: yes
    keep_profiles: no
    output_path: ~/.config/beets/xtraction_data
    essentia_extractor: /usr/bin/essentia_streaming_extractor_music
    extractor_profile:
        outputFormat: json
        outputFrames: 0
        lowlevel:
            frameSize: 2048
            hopSize: 1024
        chromaprint:
            compute: 1
            duration: 20.0
        highlevel:
            compute: 1
            svm_models:
                - /usr/share/musicplayerplus/svm_models/danceability.history
                - /usr/share/musicplayerplus/svm_models/gender.history
                - /usr/share/musicplayerplus/svm_models/genre_dortmund.history
                - /usr/share/musicplayerplus/svm_models/genre_electronic.history
                - /usr/share/musicplayerplus/svm_models/genre_rosamerica.history
                - /usr/share/musicplayerplus/svm_models/genre_tzanetakis.history
                - /usr/share/musicplayerplus/svm_models/ismir04_rhythm.history
                - /usr/share/musicplayerplus/svm_models/mood_acoustic.history
                - /usr/share/musicplayerplus/svm_models/mood_aggressive.history
                - /usr/share/musicplayerplus/svm_models/mood_electronic.history
                - /usr/share/musicplayerplus/svm_models/mood_happy.history
                - /usr/share/musicplayerplus/svm_models/mood_party.history
                - /usr/share/musicplayerplus/svm_models/mood_relaxed.history
                - /usr/share/musicplayerplus/svm_models/mood_sad.history
                - /usr/share/musicplayerplus/svm_models/moods_mirex.history
                - /usr/share/musicplayerplus/svm_models/timbre.history
                - /usr/share/musicplayerplus/svm_models/tonal_atonal.history
                - /usr/share/musicplayerplus/svm_models/voice_instrumental.history
    low_level_targets:
        bpm:
            path: rhythm.bpm
            type: integer
            required: yes
        danceability:
            path: rhythm.danceability
            type: float
        beats_count:
            path: rhythm.beats_count
            type: integer
        average_loudness:
            path: lowlevel.average_loudness
            type: float
            required: yes
    high_level_targets:
        danceable:
            path: highlevel.danceability.all.danceable
            type: float
            required: yes
        gender:
            path: highlevel.gender.value
            type: string
            required: yes
        is_male:
            path: highlevel.gender.all.male
            type: float
        is_female:
            path: highlevel.gender.all.female
            type: float
        genre_rosamerica:
            path: highlevel.genre_rosamerica.value
            type: string
            required: yes
        voice_instrumental:
            path: highlevel.voice_instrumental.value
            type: string
            required: yes
        is_voice:
            path: highlevel.voice_instrumental.all.voice
            type: float
        is_instrumental:
            path: highlevel.voice_instrumental.all.instrumental
            type: float
        mood_acoustic:
            path: highlevel.mood_acoustic.all.acoustic
            type: float
            required: yes
        mood_aggressive:
            path: highlevel.mood_aggressive.all.aggressive
            type: float
            required: yes
        mood_electronic:
            path: highlevel.mood_electronic.all.electronic
            type: float
            required: yes
        mood_happy:
            path: highlevel.mood_happy.all.happy
            type: float
            required: yes
        mood_sad:
            path: highlevel.mood_sad.all.sad
            type: float
            required: yes
        mood_party:
            path: highlevel.mood_party.all.party
            type: float
            required: yes
        mood_relaxed:
            path: highlevel.mood_relaxed.all.relaxed
            type: float
            required: yes
        mood_mirex:
            path: highlevel.moods_mirex.value
            type: string
            required: yes
        mood_mirex_cluster_1:
            path: highlevel.moods_mirex.all.Cluster1
            type: float
        mood_mirex_cluster_2:
            path: highlevel.moods_mirex.all.Cluster2
            type: float
        mood_mirex_cluster_3:
            path: highlevel.moods_mirex.all.Cluster3
            type: float
        mood_mirex_cluster_4:
            path: highlevel.moods_mirex.all.Cluster4
            type: float
        mood_mirex_cluster_5:
            path: highlevel.moods_mirex.all.Cluster5
            type: float

ui:
    color: yes
    colors:
        text_success: green
        text_warning: yellow
        text_error: red
        text_highlight: red
        text_highlight_minor: lightgray
        action_default: turquoise
        action: blue
hook:
    hooks: [{event: import_begin, command: /usr/share/musicplayerplus/scripts/beet_log.sh Import begin}, {event: import, command: /usr/share/musicplayerplus/scripts/beet_log.sh Import end}]

edit:
    albumfields: album albumartist
    itemfields: track title artist album
    ignore_fields: id path

describe:
    auto: no

missing:
    count: no
    total: no
    album: no
```

My plugin version (output of `beet xtractor -v`) is:

```text
xtractor: Xtractor(beets-xtractor) plugin for Beets: v0.3.1
```

**Comment by JOJ0** [comment]
> ### Problem
> I cannot seem to get the `xtractor` plugin to write the bpm metadata tag to the media file itself. If the file has no `bpm` tag then none is added and if it has an existing bpm tag that is not updated. Is the `write` configuration and command-line option implemented? If so, what am I doing wrong?
> 
> The calculated bpm value is correctly being added to the Beets library for the media but not to the media.

The actual writing to metadata fields of mediafiles is not implemented yet but some examples on what needs to be done is prepared here: https://github.com/adamjakab/BeetsPluginXtractor/blob/devel/beetsplug/xtractor/__init__.py#L23-L32

If you find the time and would like to help implement it you could file a PR. Funnily enough I was working on a PR that was supposed to did just that with the beets acousticbrainz plugin. I closed it since AB is discontinued soon. The PR is actually pretty much finished and includes quite some fields that this plugin uses as well. You could just copy it over from there. I won't find the time currently to do it but could help with reviewing your PR since Adam kindly gave me access to this repo :-)

This is the PR, have a look at the diff and you'll find what we are looking for :-)) https://github.com/beetbox/beets/pull/4287/files#diff-c24f6111316e743263bb7d9664f23728373413c4c99fc65345741d7c78f145e2R152

And certainly some imports would be required: https://github.com/beetbox/beets/pull/4287/files#diff-c24f6111316e743263bb7d9664f23728373413c4c99fc65345741d7c78f145e2R23-R24

Hope that helps for now and motivates you to work on this plugin together :-) I hope I'll find more time for it myself in the future!

**Comment by doctorfree** [comment]
Thanks for the pointers and references. I'm preparing a new release of [MusicPlayerPlus](https://github.com/doctorfree/MusicPlayerPlus#readme) so will not have time to work on metadata writing for a week or two. However, if all goes well with the release I will take a look and see if I can help. Note that the new release of MusicPlayerPlus will include a pre-compiled and packaged `essentia_streaming_extractor_music` binary with Gaia2 support. This seems to me to be the biggest hurdle for folks wishing to use the xtractor plugin. I look forward to any collaboration we might undertake.

**Comment by JOJ0** [comment]
But the be honest and to answer your question straight, writing the field bpm to the file, even without all the other essentia fields, should be possible because bpm is a default field, beets (actually the mediafile package in charge for that ) already is aware of. So if something like this is implemented, actually it should be possible to write it to file: https://github.com/beetbox/beets/blob/master/beetsplug/acousticbrainz.py#L232-L233

I had a look in the xtractor code and it seems to be here: https://github.com/adamjakab/BeetsPluginXtractor/blob/devel/beetsplug/xtractor/command.py#L192-L193

I don't use th xtractor myself by now and the only thing I can suggest to further debug your issue is to put some print() statements around this method to find out if it's ever called. 

Get back with your findings here, I'm curious what's wrong. We'll figure it out! :-)

**Comment by JOJ0** [comment]
Well, the call of this method was not far from the definition. It's commented out LOL: https://github.com/adamjakab/BeetsPluginXtractor/blob/devel/beetsplug/xtractor/command.py#L182

Remove the comment and give it a try, but make sure to have a backup to the file you are experimenting with. I think every field known to beets, should then be written. I think bpm and initial_key should work.

**Comment by doctorfree** [comment]
I performed a quick test this morning simply uncommenting the call to `self._run_write_to_item(item)` in command.py. This resulted in a successful write of bpm value to an mp3 file as well as the Beets library. This was only an initial test but looks good and is a trivial change. I did not see any change to `initial_key` but it may be that I need to add a low-level configuration to the xtractor section of the Beets config.yaml - I have not yet had time to look into initial_key.

Thanks for the very easy fix for bpm write to file
